src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectState.cs

```diff
@@ -274,26 +274,36 @@ public ProjectState WithChangedHostDocument(HostDocument hostDocument, TextLoade
 
     public ProjectState WithHostProjectAndWorkspaceState(HostProject hostProject, ProjectWorkspaceState projectWorkspaceState)
     {
-        if (HostProject.Configuration.Equals(hostProject.Configuration) &&
-            HostProject.RootNamespace == hostProject.RootNamespace &&
-            ProjectWorkspaceState.Equals(projectWorkspaceState))
+        var configUnchanged = (HostProject.Configuration.Equals(hostProject.Configuration) &&
+            HostProject.RootNamespace == hostProject.RootNamespace);
+
+        if (configUnchanged && ProjectWorkspaceState.Equals(projectWorkspaceState))
         {
             return this;
         }
 
-        var documents = Documents.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.WithProjectChange(), FilePathNormalizingComparer.Instance);
+        var documents = Documents.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.WithProjectChange(cacheComputedState: configUnchanged), FilePathNormalizingComparer.Instance);
+
+        var importsToRelatedDocuments = configUnchanged
+            ? ImportsToRelatedDocuments
+            : ComputeImportsToRelatedDocuments(documents);
 
-        // If the host project has changed then we need to recompute the imports map
-        var importsToRelatedDocuments = s_emptyImportsToRelatedDocuments;
+        var state = new ProjectState(this, numberOfDocumentsMayHaveChanged: !configUnchanged, hostProject, projectWorkspaceState, documents, importsToRelatedDocuments);
```

Comment:
It's a bit hidden, so I'll call it out here. Where I said in the PR description that updating the `DocumentCollectionVersion` was the main issue, its this `numberOfDocumentsMayHaveChanged` parameter.

In my original PR I equated any project change to a possible document change, but tag helpers changing is definitely not that.